File: openc3/lib/openc3/utilities/bucket.rb
# encoding: ascii-8bit

# Copyright 2026 OpenC3, Inc.
# All Rights Reserved.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE.md for more details.
#
# This file may also be used under the terms of a commercial license
# if purchased from OpenC3, Inc.

ENV['OPENC3_CLOUD'] ||= 'local'

# Interface class implemented by each cloud provider: AWS, GCS, Azure
module OpenC3
  class Bucket
    # Raised when the underlying bucket does not exist
    class NotFound < RuntimeError
    end

    def self.getClient
      raise 'OPENC3_CLOUD environment variable is required' unless ENV['OPENC3_CLOUD']
      # Base is AwsBucket which works with S3-compatible storage (versitygw), Enterprise implements additional
      bucket_class = ENV['OPENC3_CLOUD'].capitalize + 'Bucket'
      klass = OpenC3.require_class('openc3/utilities/'+bucket_class.class_name_to_filename)
      klass.new
    end

    def initialize
      # Setup the client instance
    end

    def create(bucket)
      raise NotImplementedError, "#{self.class} has not implemented method '#{__method__}'"
    end

    def ensure_public(bucket)
      raise NotImplementedError, "#{self.class} has not implemented method '#{__method__}'"
    end

    # Apply bucket policies for ScriptRunner user (only applicable for local mode)
    def ensure_scriptrunner_policy(config_bucket, logs_bucket)
      # No-op by default, implemented by AwsBucket for local mode
    end

    def exist?(bucket)
      raise NotImplementedError, "#{self.class} has not implemented method '#{__method__}'"
    end

    def delete(bucket)
      raise NotImplementedError, "#{self.class} has not implemented method '#{__method__}'"
    end

    def get_object(bucket:, key:, path: nil, range: nil)
      raise NotImplementedError, "#{self.class} has not implemented method '#{__method__}'"
    end

    def list_objects(bucket:, prefix: nil, max_request: nil, max_total: nil)
      raise NotImplementedError, "#{self.class} has not implemented method '#{__method__}'"
    end

    def list_files(bucket:, path:, only_directories: false, metadata: false)
      raise NotImplementedError, "#{self.class} has not implemented method '#{__method__}'"
    end

    def put_object(bucket:, key:, body:, content_type: nil, cache_control: nil, metadata: nil)
      raise NotImplementedError, "#{self.class} has not implemented method '#{__method__}'"
    end

    def check_object(bucket:, key:, retries: true)
      raise NotImplementedError, "#{self.class} has not implemented method '#{__method__}'"
    end

    def delete_object(bucket:, key:)
      raise NotImplementedError, "#{self.class} has not implemented method '#{__method__}'"
    end

    def delete_objects(bucket:, keys:)
      raise NotImplementedError, "#{self.class} has not implemented method '#{__method__}'"
    end

    def presigned_request(bucket:, key:, method:, internal: true)
      raise NotImplementedError, "#{self.class} has not implemented method '#{__method__}'"
    end
  end
end

File: openc3/python/openc3/utilities/bucket.py
# Copyright 2026 OpenC3, Inc.
# All Rights Reserved.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE.md for more details.
#
# This file may also be used under the terms of a commercial license
# if purchased from OpenC3, Inc.

import importlib
import inspect
import warnings

from openc3.environment import OPENC3_CLOUD


# Interface class implemented by each cloud provider: AWS, GCS, Azure
class Bucket:
    # Raised when the underlying bucket does not exist
    class NotFoundError(Exception):
        pass

    @classmethod
    def get_client(cls):
        if not OPENC3_CLOUD:
            raise RuntimeError("OPENC3_CLOUD environment variable is required")
        # Base is AwsBucket which works with S3-compatible storage (versitygw), Enterprise implements additional
        bucket_class = OPENC3_CLOUD.capitalize() + "Bucket"
        my_module = None
        try:
            my_module = importlib.import_module("." + OPENC3_CLOUD.lower() + "_bucket", "openc3.utilities")
        # If the file doesn't exist try the Enterprise module
        except ModuleNotFoundError:
            my_module = importlib.import_module("." + OPENC3_CLOUD.lower() + "_bucket", "openc3enterprise.utilities")
        return getattr(my_module, bucket_class)()

    @classmethod
    def getClient(cls):  # noqa: N802
        """Deprecated: Use get_client() instead. This camelCase method is provided for backwards compatibility."""
        warnings.warn(
            "Bucket.getClient() is deprecated, use Bucket.get_client() instead",
            DeprecationWarning,
            stacklevel=2
        )
        return cls.get_client()

    def create(self, bucket):
        raise NotImplementedError(
            f"{self.__class__.__name__} has not implemented method '{inspect.currentframe().f_code.co_name}'"
        )

    def ensure_public(self, bucket):
        raise NotImplementedError(
            f"{self.__class__.__name__} has not implemented method '{inspect.currentframe().f_code.co_name}'"
        )

    def exist(self, bucket):
        raise NotImplementedError(
            f"{self.__class__.__name__} has not implemented method '{inspect.currentframe().f_code.co_name}'"
        )

    def delete(self, bucket):
        raise NotImplementedError(
            f"{self.__class__.__name__} has not implemented method '{inspect.currentframe().f_code.co_name}'"
        )

    def get_object(self, bucket, key, path=None, range=""):
        raise NotImplementedError(
            f"{self.__class__.__name__} has not implemented method '{inspect.currentframe().f_code.co_name}'"
        )

    def list_objects(self, bucket, prefix=None, max_request=None, max_total=None):
        raise NotImplementedError(
            f"{self.__class__.__name__} has not implemented method '{inspect.currentframe().f_code.co_name}'"
        )

    def list_files(self, bucket, path, only_directories=False, metadata=False):
        raise NotImplementedError(
            f"{self.__class__.__name__} has not implemented method '{inspect.currentframe().f_code.co_name}'"
        )

    def put_object(self, bucket, key, body, content_type=None, cache_control=None, metadata=None):
        raise NotImplementedError(
            f"{self.__class__.__name__} has not implemented method '{inspect.currentframe().f_code.co_name}'"
        )

    def check_object(self, bucket, key, retries=True):
        raise NotImplementedError(
            f"{self.__class__.__name__} has not implemented method '{inspect.currentframe().f_code.co_name}'"
        )

    def delete_object(self, bucket, key):
        raise NotImplementedError(
            f"{self.__class__.__name__} has not implemented method '{inspect.currentframe().f_code.co_name}'"
        )

    def delete_objects(self, bucket, keys):
        raise NotImplementedError(
            f"{self.__class__.__name__} has not implemented method '{inspect.currentframe().f_code.co_name}'"
        )

    def presigned_request(self, bucket, key, method, internal=True):
        raise NotImplementedError(
            f"{self.__class__.__name__} has not implemented method '{inspect.currentframe().f_code.co_name}'"
        )
